File: Napakalaki_Java/src/napakalaki/Player.java
/*
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
*/
package napakalaki;

import java.util.ArrayList;
import java.util.Random;

public class Player {
    //Atributos
    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal = true;
    
    private Player enemy;
    private ArrayList<Treasure> hiddenTreasures = new ArrayList();
    private ArrayList<Treasure> visibleTreasures = new ArrayList();
    private BadConsequence pendingBadConsequence;
    
    //Métodos
    public Player(String name) {
        this.name = name;
        this.level = 1;
    }
    
    //Devuelve el nombre del jugador.
    public String getName() {
        return this.name;
    }
    
    //Devuelve la vida al jugador, modificando el atributo correspondiente.
    private void bringToLife() {
        this.dead = false;
    }
    
    //Devuelve el nivel de combate del jugador, que viene dado por su nivel más los bonus
    //que le proporcionan los tesoros que tenga equipados, según las reglas del juego
    private int getCombatLevel() {
        int combatlevel = this.level;
        for (Treasure t : visibleTreasures)
            combatlevel += t.getBonus();
        return combatlevel;
    }
    
    //Incrementa el nivel del jugador en i niveles, teniendo en cuenta las reglas del juego.
    private void incrementLevels(int i) {
        this.level += i;
    }
    
    //Decrementa el nivel del jugador en i niveles, teniendo en cuenta las reglas del juego.
    private void decrementLeveles(int i) {
        this.level -= i;
    }
    
    //Asigna el mal rollo al jugador, dándole valor a su atributo pendingBadConsequence.
    private void setPendingBadConsequence(BadConsequence b) {
        this.pendingBadConsequence = b;
    }
    
    private void applyPrize(Monster m) {
        int nLevels = m.getLevelsGained();
        int l = nLevels;
        this.incrementLevels(l);
        int nTreasures = m.getTreasuresGained();
        
        if(nTreasures > 0) {
            CardDealer dealer = CardDealer.getInstance();
            
            for(int i=0; i<nTreasures; i++) {
                Treasure treasure = dealer.nextTreasure();
                this.hiddenTreasures.add(treasure);
            }
        }
    }
    
    private void applyBadConsequence(Monster m) {
        BadConsequence badConsequence = m.getBadConsequence();
        int l = badConsequence.getLevels();
        this.decrementLeveles(l);
        ArrayList<Treasure> v = this.visibleTreasures, h = this.hiddenTreasures;
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureList(v, h);
        this.setPendingBadConsequence(pendingBad);
    }
    
    //Antes de cada combate, y antes de conocer el monstruo con el que se enfrentará, el jugador puede
    //equiparse los tesoros que desee siempre que cumpla las reglas sobre la cantidad y tipos de tesoros
    //que se pueden tener equipados. Equipar un tesoro implica quitarlo del conjunto de los ocultos y
    //pasarlo al de los equipados.
    private boolean canMakeTreasureVisible(Treasure t) {
        int ntreasuresonehand = 0;
        boolean treasurebothhands = false, output = false, mismotipo=false;
        if (this.visibleTreasures.size()<4){
            for(Treasure tesoro : this.visibleTreasures) {
                if (tesoro.getType() == TreasureKind.ONEHAND)
                    ntreasuresonehand++;
                else if (tesoro.getType() == TreasureKind.BOTHHANDS)
                    treasurebothhands = true;
                else if (t.getType() == tesoro.getType())     
                    mismotipo = true;
            }
            if (ntreasuresonehand < 3 && t.getType() == TreasureKind.ONEHAND)
                output = true;
            else if (ntreasuresonehand == 0 && treasurebothhands == true)
                output = true;
            else if (mismotipo == false)
                output = true;
        }
        
        return output;
    }
    
    //Devuelve el número de tesoros visibles de tipo tKind que tiene el jugador.
    private int howManyVisibleTreasures(TreasureKind tKind) {
        int counter = 0;
        for (Treasure treasure : this.visibleTreasures)
            if (treasure.getType() == tKind)
                counter++;
        return counter;
    }
    
    //Cambia el estado de jugador a muerto, modificando el correspondiente atributo.
    //Esto ocurre cuando el jugador, por algún motivo, ha perdido todos sus tesoros.
    private void dieIfNoTreasures() {
        if (hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            this.dead = true;
    }
    
    //Devuelve true si el jugador está muerto, false en caso contrario.
    public boolean isDead() {
        return this.dead;
    }
    
    public ArrayList<Treasure> getHiddenTreasures() {
        return this.hiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures() {
        return this.visibleTreasures;
    }
    
    public CombatResult combat(Monster m) {
        int myLevel = this.getCombatLevel();
        int monsterLevel = m.getCombatLevel();
        CombatResult combatResult;
        if(myLevel > monsterLevel) {
            this.applyPrize(m);
            
            if(this.level >= this.MAXLEVEL){
                combatResult = CombatResult.WINGAME;
            }
            else combatResult = CombatResult.WIN;
        }
        else {
            this.applyBadConsequence(m);
            
            combatResult = CombatResult.LOSE;
        }
        
        return combatResult;
    }
    
    public void makeTreasureVisible(Treasure t) {
        boolean canI = this.canMakeTreasureVisible(t);
        if(canI) {
            this.visibleTreasures.add(t);
            this.hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t) {
        this.visibleTreasures.remove(t);
        if((this.pendingBadConsequence != null) && (!this.pendingBadConsequence.isEmpty())){
            this.pendingBadConsequence.substractVisibleTreasure(t);
        }
        this.dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t) {
        this.visibleTreasures.remove(t);
        if((this.pendingBadConsequence != null) && (!this.pendingBadConsequence.isEmpty())){
            this.pendingBadConsequence.substractHiddenTreasure(t);
        }
        this.dieIfNoTreasures();
    }
    
    //Devuelve true cuando el jugador no tiene ningún mal rollo que cumplir y no tiene
    //más de 4 tesoros ocultos, y false en caso contrario.
    public boolean validState() {
        boolean v = false;
        System.out.println("pendingBadConsequence: " + this.pendingBadConsequence.isEmpty());
        System.out.println("hiddenTreasures.size: " + this.hiddenTreasures.size());
        if (this.pendingBadConsequence == null || (this.pendingBadConsequence.isEmpty() && this.hiddenTreasures.size() <= 4))
            v = true;
        System.out.println("v: " + v);
        return v;
    }
    
    public void initTreasures() {
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        this.bringToLife();
        Treasure treasure = dealer.nextTreasure();
        this.hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        if(number > 1) {
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
        if(number == 6) {
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
    }
    
    //Devuelve el nivel del jugador.
    public int getLevels() {
        return this.level;
    }
    
    public Treasure stealTreasure() {
        boolean canI = this.canISteal();
        Treasure treasure = null;
        if (canI) {
            boolean canYou = this.enemy.canYouGiveMeATreasure();
            if (canYou) {
                treasure = this.enemy.giveMeATreasure();
                this.hiddenTreasures.add(treasure);
                this.haveStolen();
            }
        }
        return treasure;
    }
    
    //Asigna valor al atributo que referencia al enemigo del jugador.
    public void setEnemy(Player enemy) {
        this.enemy = enemy;
    }
    
    //Devuelve un tesoro elegido al azar de entre los tesoros ocultos del jugador.
    private Treasure giveMeATreasure() {
        Random random = new Random();
        int index = random.nextInt(this.hiddenTreasures.size());
        return this.hiddenTreasures.get(index);
    }
    
    //Devuelve true si el jugador no ha robado ningún tesoro a su enemigo y false en
    //caso contrario.
    public boolean canISteal() {
        return this.canISteal;
    }
    
    //Devuelve true si el jugador tiene tesoros para ser robados por otro jugador y false
    //en caso contrario.
    private boolean canYouGiveMeATreasure() {
        boolean youcan = false;
        if (!this.hiddenTreasures.isEmpty())
            youcan = true;
        return youcan;
    }
    
    //Cambia el atributo canISteal a false cuando el jugador roba un tesoro.
    private void haveStolen() {
        this.canISteal = false;
    }
    
    public void discardAllTreasures() {
        for(Treasure treasure : this.visibleTreasures) {
            this.discardVisibleTreasure(treasure);
        }
        for(Treasure treasure : this.hiddenTreasures) {
            this.discardHiddenTreasure(treasure);
        }
    }
    
    @Override
    public String toString() {
        return this.name;
    }
}

File: Napakalaki_Ruby/lib/player.rb
# encoding: utf-8

=begin
********************************************************************
**       _   _                   _         _       _    _         **
**      | \ | | __ _ _ __   __ _| | ____ _| | __ _| | _(_)        **
**      |  \| |/ _` | '_ \ / _` | |/ / _` | |/ _` | |/ / |        **
**      | |\  | (_| | |_) | (_| |   < (_| | | (_| |   <| |        **
**      |_| \_|\__,_| .__/ \__,_|_|\_\__,_|_|\__,_|_|\_\_|        **
**                  |_|                                           **
** Creado por Javier Bolívar Valverde <javibolivar@correo.ugr.es> **
** Programación y Diseño Orientado a Objetos                      **
** Grado en Ingeniería Informática (Universidad de Granada)       **
********************************************************************
=end

require_relative 'dice'
require_relative 'bad_consequence'
require_relative 'combat_result'

module Napakalaki
  class Player

    @@MAXLEVEL = 10

    def initialize(name)
      @name = name
      @level = 1
      @dead = true
      @can_i_steal = true
      @hidden_treasures = Array.new
      @visible_treasures = Array.new
      @pending_bad_consequence = nil
      @enemy = nil
    end

    attr_reader :name, :level, :dead, :can_i_steal, :hidden_treasures, :visible_treasures
    attr_writer :pending_bad_consequence, :enemy

    private
    #Devuelve la vida al jugador, modificando el atributo correspondiente.
    def bring_to_life
      @dead = false
    end

    #Devuelve el nivel de combate del jugador, que viene dado por su nivel más los bonus
    #que le proporcionan los tesoros que tenga equipados, según las reglas del juego.
    def get_combat_level
      combatlevel = @level

      visible_treasures.each do |t|
        combatlevel += t.bonus
      end
      return combatlevel
    end

    #Incrementa el nivel del jugador en i niveles, teniendo en cuenta las reglas del juego.
    def increment_levels(i)
      @level += i
    end

    #Decrementa el nivel del jugador en i niveles, teniendo en cuenta las reglas del juego.
    def decrement_levels(i)
      @level -= i
    end

    def apply_prize(m)
      n_levels = m.get_levels_gained
      l = n_levels
      increment_levels(l)
      n_treasures = m.get_treasures_gained

      if(n_treasures > 0)
        dealer = CardDealer.instance

        for counter in 0..n_treasures
          treasure = dealer.next_treasure
          @hidden_treasures << treasure
        end
      end
    end

    def apply_bad_consequence(m)
      b = m.bad_consequence
      l = b.levels
      decrement_levels(l)
      v = @visible_treasures
      h = @hidden_treasures
      pending_bad = b.adjust_to_fit_treasure_lists(v, h)
      @pending_bad_consequence = pending_bad
    end

    def can_make_treasure_visible(t)
      n_treasures_one_hand = 0
      treasure_both_hands = false, output = false, mismo_tipo = false
      if(@visible_treasures.size < 4)
        @visible_treasures.each do |tesoro|
          if (tesoro.type == TreasureKind::ONEHAND)
            n_treasures_one_hand+=1
          elsif (tesoro.type == TreasureKind::BOTHHANDS)
            treasure_both_hands = true
          elsif (t.type == tesoro.type)
            mismo_tipo = true
          end
        end

        if (n_treasures_one_hand < 3 && t.type == TreasureKind::ONEHAND)
          output = true
        elsif (n_treasures_one_hand == 0 && treasure_both_hands == true)
          output = true
        elsif (mismo_tipo == false)
          output = true
        end
      end
      return output
    end

    #Devuelve el número de tesoros visibles de tipo tKind que tiene el jugador.
    def how_many_visible_treasures(tKind)
      counter = 0
      @visible_treasures.each do |t|
        counter += 1 if (t == tKind)
      end
      return counter
    end

    #Cambia el estado de jugador a muerto, modificando el correspondiente atributo.
    #Esto ocurre cuando el jugador, por algún motivo, ha perdido todos sus tesoros.
    def die_if_no_treasures
      @dead = true if (@hidden_treasures.empty? && @visible_treasures.empty?)
    end

    #Cambia el atributo canISteal a false cuando el jugador roba un tesoro.
    def have_stolen
      @can_i_steal = false
    end

    public  
    def combat(m)
      my_level = get_combat_level
      monster_level = m.combat_level
      if(my_level > monster_level)
          apply_prize(m)

          if(@level >= @@MAXLEVEL)
            combat_result = CombatResult::WINGAME
          else
            combat_result = CombatResult::WIN
          end
      else
        apply_bad_consequence(m)
        combat_result = CombatResult::LOSE
      end
    end

    def make_treasure_visible(t)
      can_i = can_make_treasure_visible(t)
      if (can_i)
        @visible_treasures << t
        @hidden_treasures.delete(t)
      end
    end

    def discard_visible_treasure(t)
      @visible_treasures.delete(t)
      if((@pending_bad_consequence != nil) && (!@pending_bad_consequence.is_empty))
        @pending_bad_consequence.substract_visible_treasure(t)
      end
      die_if_no_treasures
    end

    def discard_hidden_treasure(t)
      @visible_treasures.delete(t)
      if((@pending_bad_consequence != nil) && (!@pending_bad_consequence.is_empty))
        @pending_bad_consequence.substract_hidden_treasure(t)
      end
      die_if_no_treasures
    end

    #Devuelve true cuando el jugador no tiene ningún mal rollo que cumplir y no tiene
    #más de 4 tesoros ocultos, y false en caso contrario.
    def valid_state
      v = false
      v = true if (@pending_bad_consequence.is_empty && @hidden_treasures.size <= 4)
      return v
    end

    def init_treasures
      dealer = CardDealer.instance
      dice = Dice.instance
      bring_to_life
      treasure = dealer.next_treasure
      @hidden_treasures << treasure
      number = dice.next_number
      if(number > 1)
        treasure = dealer.next_treasure
        @hidden_treasures << treasure
      end
      if(number == 6)
        treasure = dealer.next_treasure
        @hidden_treasures << treasure
      end
    end
    
    #Devuelve true si el jugador tiene tesoros para ser robados por otro jugador y false
    #en caso contrario.
    def can_you_give_me_a_treasure
      can = false
      if(!@visible_treasures.empty? || !@hidden_treasures.empty?)
        can = true
      end
      return can
    end
    
    def give_me_a_treasure
      index = rand(@hidden_treasures.size)
      return @hidden_treasures.at(index)
    end

    def steal_treasure
      can_i = @can_i_steal
      if (can_i)
        can_you = @enemy.can_you_give_me_a_treasure
        if(can_you)
          treasure = @enemy.give_me_a_treasure
          @hidden_treasures << treasure
          have_stolen
        end
      end
      treasure
    end

    def discard_all_treasures
      @visible_treasures.each do |treasure|
        discard_visible_treasure(treasure)
      end
      @hidden_treasures.each do |treasure|
        discard_hidden_treasure(treasure)
      end
    end
    
    def to_s
      @name
    end
  end
end
